numpy.inner return wrong inaccurate value sometimes...

After updating bumpy from 1.9.2-py27_0 to 1.10.1-py27_0. One of my class method returned a different and thus wrong result.
I localised the error to be from the numpy.inner fonction. What is weird is that when I use the method within another class method declared in a different file, it works! 

When I use the class after
 
if __name__ == '__main__': 

in the same file the class is written to test it it does not work even though the input data is exactly the same in each case. When I declare the class within a Jupyter notebook, the same problem occurs, a wrong output is given by numpy.inner.

('Numpy version', '1.10.1')
((50,), (50,), -0.48859534954264638, 0.47975144411505222, -41.144742080228653, 39.598697722152565)
(-28.278582107027624, 302.06900546026594, 302.06900546026594)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-23-2129c54d7dc0> in <module>()
     26 
     27 if __name__ == '__main__':
---> 28     doit()

<ipython-input-23-2129c54d7dc0> in doit()
     22     print(c, c2, c3)
     23 
---> 24     np.testing.assert_allclose(c, c2)
     25     np.testing.assert_allclose(c, c3)
     26 

/Users/olivier/anaconda/envs/testnumpy/lib/python2.7/site-packages/numpy/testing/utils.pyc in assert_allclose(actual, desired, rtol, atol, equal_nan, err_msg, verbose)
   1345     header = 'Not equal to tolerance rtol=%g, atol=%g' % (rtol, atol)
   1346     assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
-> 1347                          verbose=verbose, header=header)
   1348 
   1349 def assert_array_almost_equal_nulp(x, y, nulp=1):

/Users/olivier/anaconda/envs/testnumpy/lib/python2.7/site-packages/numpy/testing/utils.pyc in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision)
    706                                 names=('x', 'y'), precision=precision)
    707             if not cond:
--> 708                 raise AssertionError(msg)
    709     except ValueError:
    710         import traceback

AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(mismatch 100.0%)
 x: array(-28.278582107027624)
 y: array(302.06900546026594)

> On 21 Oct 2015, at 13:28, Pauli Virtanen <notifications@github.com> wrote:
> 
> import numpy as np
> 
> def doit():
>     np.random.seed(1234)
>     X = np.random.rand(50, 1000)
>     Y = np.random.rand(50, 3)
>     X = X - X.mean(axis=0)
>     Y = Y - Y.mean(axis=0)
> 
>     y = Y[:,0]
> 
>     XX = X.dot(X.T)
>     tn = XX.dot(y)
> 
>     print("Numpy version", np.__version__)
>     print(y.shape, tn.shape, y.min(), y.max(), tn.min(), tn.max())
> 
>     c = np.inner(y, tn) 
>     c2 = np.inner(y.copy(), tn) 
>     c3 = (y*tn).sum()
> 
>     print(c, c2, c3)
> 
>     np.testing.assert_allclose(c, c2)
>     np.testing.assert_allclose(c, c3)
> 
> if __name__ == '__main__':
>     doit()

--- a/numpy/core/src/multiarray/cblasfuncs.c
+++ b/numpy/core/src/multiarray/cblasfuncs.c
@@ -674,8 +674,8 @@ fail:
  *
  * This is for use by PyArray_InnerProduct. It is assumed on entry that the
  * arrays ap1 and ap2 have a common data type given by typenum that is
- * float, double, cfloat, or cdouble and have dimension <= 2, and have the
- * contiguous flag set. The * __numpy_ufunc__ nonsense is also assumed to
+ * float, double, cfloat, or cdouble and have dimension <= 2.
+ * The * __numpy_ufunc__ nonsense is also assumed to
  * have been taken care of.
  */
 
@@ -688,6 +688,24 @@ cblas_innerproduct(int typenum, PyArrayObject *ap1, PyArrayObject *ap2)
     PyArrayObject *ret = NULL;
     npy_intp dimensions[NPY_MAXDIMS];
     PyTypeObject *subtype;
+
+    /* assure contiguous arrays */
+    if (!PyArray_IS_C_CONTIGUOUS(ap1)) {
+        PyObject *op1 = PyArray_NewCopy(ap1, NPY_CORDER);
+        Py_DECREF(ap1);
+        ap1 = (PyArrayObject *)op1;
+        if (ap1 == NULL) {
+            goto fail;
+        }
+    }
+    if (!PyArray_IS_C_CONTIGUOUS(ap2)) {
+        PyObject *op2 = PyArray_NewCopy(ap2, NPY_CORDER);
+        Py_DECREF(ap2);
+        ap2 = (PyArrayObject *)op2;
+        if (ap2 == NULL) {
+          goto fail;
+        }
+    }
 
     if (PyArray_NDIM(ap1) == 0 || PyArray_NDIM(ap2) == 0) {
         /* One of ap1 or ap2 is a scalar */

--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -4728,6 +4728,22 @@ class TestInner(TestCase):
         p = np.inner(a, a)
         assert_almost_equal(p, 0, decimal=14)
 
+    def test_inner_product_with_various_contiguities(self):
+        # github issue 6532
+        for dt in np.typecodes['AllInteger'] + np.typecodes['AllFloat'] + '?':
+            # check an inner product involving a matrix transpose
+            A = np.array([[1, 2], [3, 4]], dtype=dt)
+            B = np.array([[1, 3], [2, 4]], dtype=dt)
+            C = np.array([1, 1], dtype=dt)
+            desired = np.array([4, 6], dtype=dt)
+            assert_equal(np.inner(A.T, C), desired)
+            assert_equal(np.inner(B, C), desired)
+            # check an inner product involving an aliased and reversed view
+            a = np.arange(5).astype(dt)
+            b = a[::-1]
+            desired = np.array(10, dtype=dt).item()
+            assert_equal(np.inner(b, a), desired)
+
 
 class TestSummarization(TestCase):
     def test_1d(self):

